Upgrade Playwright

TanayParikh opened the pull request:
This is used within the Blazor Template Tests.

Contributes towards https://github.com/dotnet/aspnetcore/issues/30882.


Note, the goal of this PR is just to get the underlying Playwright upgraded from the pre-release version, to the latest stable release. The underlying tests are unchanged by this PR.

TanayParikh committed:
Upgrade `playwright-sharp-tool`

https://www.nuget.org/packages/playwright-sharp-tool

Contributes towards https://github.com/dotnet/aspnetcore/issues/30882.

TanayParikh commented:
Converting to draft. Nuget hasn't been updated in 8 months and it's stuck on 0.192.0 (https://www.nuget.org/packages/playwright-sharp-tool), while the GitHub repo is on 1.16.1 (https://github.com/microsoft/playwright-dotnet/tags). Investigating upgrade.

TanayParikh commented on .config/dotnet-tools.json:
```diff
@@ -9,7 +9,7 @@
     },
     "playwright-sharp-tool": {
-      "version": "0.170.2",
+      "version": "0.192.0",
```
```suggestion
      "version": "1.16.1",
```

TanayParikh committed:
Update to latest v1.16.1 from GitHub

https://github.com/microsoft/playwright-dotnet/releases/tag/v1.16.1

TanayParikh committed:
Update to use Microsoft.Playwright

TanayParikh committed:
Update references to reflect `Microsoft.Playwright`

TanayParikh commented:
/azp run

TanayParikh committed:
Update dotnet-tools.json

TanayParikh commented:
/azp run

adrianwright109 commented:
@TanayParikh FYI: I just looked on the GitHub Repo [https://github.com/microsoft/playwright-dotnet](https://github.com/microsoft/playwright-dotnet) and saw that version `1.12.1` which was the next version after `0.192.0` was a major release and includes many breaking changes.  See release notes for more details: [https://github.com/microsoft/playwright-dotnet/releases/tag/v1.12.1](https://github.com/microsoft/playwright-dotnet/releases/tag/v1.12.1)

TanayParikh committed:
Upgrade System.CommandLine

TanayParikh committed:
Adapt to breaking API changes in Playwright

TanayParikh committed:
Upgrade dotnet-serve

TanayParikh committed:
Add Microsoft.Playwright.CLI

TanayParikh commented on eng/helix/content/RunTests/TestRunner.cs:
```diff
@@ -55,9 +55,6 @@ public bool SetupEnvironment()
                 var playwrightBrowsers = Environment.GetEnvironmentVariable("PLAYWRIGHT_BROWSERS_PATH");
                 Console.WriteLine($"Setting PLAYWRIGHT_BROWSERS_PATH: {playwrightBrowsers}");
                 EnvironmentVariables.Add("PLAYWRIGHT_BROWSERS_PATH", playwrightBrowsers);
-                var playrightDriver = Environment.GetEnvironmentVariable("PLAYWRIGHT_DRIVER_PATH");
```
Doesn't seem to be possible to specify driver path any longer:

https://github.com/microsoft/playwright-dotnet/blob/45de47f3f51c086278b347e09a878731d996e2e5/src/Playwright/Helpers/Paths.cs

TanayParikh committed:
Revert "Upgrade System.CommandLine"

This reverts commit cc46e9cfb7a5bd75cb6454aac07fbeb96e6040f7.

TanayParikh committed:
Merge branch 'main' into taparik/upgradePlaywrightSharpTool

TanayParikh committed:
Fix build

The build failed. Fix the build errors and run again.
C:\h\w\AA180980\w\9ED808A4\e\RunTests\RunTests.csproj : warning NU1604: Project dependency Microsoft.Playwright does not contain an inclusive lower bound. Include a lower bound in the dependency version to ensure consistent restore results.
C:\h\w\AA180980\w\9ED808A4\e\RunTests\TestRunner.cs(124,35): error CS7036: There is no argument given that corresponds to the required formal parameter 'arguments' of 'ProcessUtil.RunAsync(string, string, string?, string?, bool, IDictionary<string, string?>?, Action<string>?, Action<string>?, Action<int>?, CancellationToken)' [C:\h\w\AA180980\w\9ED808A4\e\RunTests\RunTests.csproj]

TanayParikh committed:
Remove semaphore for websocket events

TanayParikh committed:
Update playwrightSettings based on breaking changes

TanayParikh committed:
npx based playwright install

TanayParikh committed:
npm install playwright

TanayParikh commented on eng/helix/content/RunTests/TestRunner.cs:
```diff
@@ -112,8 +109,26 @@ public async Task<bool> InstallPlaywrightAsync()
         {
             try
             {
-                Console.WriteLine($"Installing Playwright to Browsers: {Environment.GetEnvironmentVariable("PLAYWRIGHT_BROWSERS_PATH")} Driver: {Environment.GetEnvironmentVariable("PLAYWRIGHT_DRIVER_PATH")}");
```
Note, the "correct" way to install browsers is using the `Microsoft.Playwright.Cli` tool, however that package only works with TFM 5 (https://github.com/microsoft/playwright-dotnet/issues/1668). Even once it's upgraded to TFM 6, our CI environments are TFM 7 (and realistically will always be ahead of the general release version).

javiercn replied:
It should be ok, the tool should be able to run in a higher TFM, it's likely also possible that previous TFMs are already installed

TanayParikh replied:
> It should be ok, the tool should be able to run in a higher TFM

Any tips on how? It would be ideal to use the tool considering there are issues with using `npm`/`npx` in the environment as well:

```
Running 'npm install playwright'
Exception installing playwright: System.ComponentModel.Win32Exception (2): An error occurred trying to start process 'npm' with working directory '/private/tmp/helix/working/AA53097D/w/B12B0925/e'. No such file or directory
   at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at RunTests.ProcessUtil.RunAsync(String filename, String arguments, String workingDirectory, String dumpDirectoryPath, Boolean throwOnError, IDictionary`2 environmentVariables, Action`1 outputDataReceived, Action`1 errorDataReceived, Action`1 onStart, CancellationToken cancellationToken) in /private/tmp/helix/working/AA53097D/w/B12B0925/e/RunTests/ProcessUtil.cs:line 167
   at RunTests.TestRunner.InstallPlaywrightAsync() in /private/tmp/helix/working/AA53097D/w/B12B0925/e/RunTests/TestRunner.cs:line 114
```


> it's likely also possible that previous TFMs are already installed

Doesn't appear to be the case from the helix logs:

```
Installing Microsoft.Playwright.CLI dotnet tool to C:\h\w\B5E109F8\w\A488088E\e
Running 'C:\h\w\B5E109F8\p\dotnet-cli/dotnet tool install Microsoft.Playwright.CLI --tool-path C:\h\w\B5E109F8\w\A488088E\e'
You can invoke the tool using the following command: playwright
Tool 'microsoft.playwright.cli' (version '1.2.0') was successfully installed.
'C:\h\w\B5E109F8\p\dotnet-cli/dotnet tool install Microsoft.Playwright.CLI --tool-path C:\h\w\B5E109F8\w\A488088E\e' completed with exit code '0'
Installing Playwright Browsers: C:\h\w\B5E109F8\w\A488088E\e\ms-playwright
Running 'playwright install'
It was not possible to find any compatible framework version
The framework 'Microsoft.NETCore.App', version '5.0.0' (x64) was not found.
  - The following frameworks were found:
      7.0.0-alpha.1.21551.1 at [C:\h\w\B5E109F8\p\dotnet-cli\shared\Microsoft.NETCore.App]
      7.0.0-alpha.1.21567.1 at [C:\h\w\B5E109F8\p\dotnet-cli\shared\Microsoft.NETCore.App]

You can resolve the problem by installing the specified framework and/or SDK.

The specified framework can be found at:
  - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=5.0.0&arch=x64&rid=win10-x64
'playwright install' completed with exit code '-2147450730'

Displaying directory contents for C:\h\w\B5E109F8\w\A488088E\e\ms-playwright:
Exception in DisplayContents: System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\h\w\B5E109F8\w\A488088E\e\ms-playwright'.
   at System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound)
   at System.IO.Enumeration.FileSystemEnumerator`1.Init()
   at System.IO.Enumeration.FileSystemEnumerable`1..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized)
   at System.IO.Enumeration.FileSystemEnumerableFactory.UserFiles(String directory, String expression, EnumerationOptions options)
   at System.IO.Directory.InternalEnumeratePaths(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options)
   at RunTests.TestRunner.DisplayContents(String path) in C:\h\w\B5E109F8\w\A488088E\e\RunTests\TestRunner.cs:line 91
```

TanayParikh replied:
cc/ @dougbu in case you have any suggestions here.

martincostello replied:
You could try using the `DOTNET_ROLL_FORWARD: Major` environment variable to force it onto running with the current runtime version. This worked for me in some personal projects ([example](https://github.com/martincostello/alexa-london-travel-site/blob/b033a2619cc5530f6302f57e6e96c200d84c600d/.github/workflows/build.yml#L21)).

TanayParikh replied:
Thanks @martincostello that did it!

TanayParikh committed:
Install node in helix env

TanayParikh committed:
Use DOTNET_ROLL_FORWARD with `dotnet` `playwright` tool

TanayParikh committed:
Cleanup

TanayParikh committed:
Update playwright tool usage

TanayParikh committed:
DOTNET_ROLL_FORWARD in TestRunner

TanayParikh committed:
Merge branch 'main' into taparik/upgradePlaywrightSharpTool

TanayParikh committed:
Specify project dir

TanayParikh committed:
Playwright install without cli

TanayParikh committed:
Update TestRunner.cs

TanayParikh committed:
Update TestRunner.cs

TanayParikh committed:
Merge branch 'main' into taparik/upgradePlaywrightSharpTool

TanayParikh committed:
Re-enable helix playwright tests

mxschmitt commented:
@TanayParikh it should now all work on .NET 6 since 1.17.3. Let us know if there are any more blockers on our side!

TanayParikh committed:
Merge branch 'main' into taparik/upgradePlaywrightSharpTool

TanayParikh committed:
Upgrade Playwright to 1.17.3, CLI to 1.2.2

TanayParikh committed:
Fix spacing

mxschmitt commented on src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:
```diff
@@ -473,21 +473,21 @@ private async Task TestBasicNavigation(string appName, IPage page, bool usesAuth
             // Asynchronously loads and displays the table of weather forecasts
             await page.WaitForSelectorAsync("table>tbody>tr");
-            Assert.Equal(5, (await page.QuerySelectorAllAsync("p+table>tbody>tr")).Count());
+            Assert.Equal(5, (await page.QuerySelectorAllAsync("p+table>tbody>tr")).Count);
```
ditto: change to locators.

mxschmitt commented on src/ProjectTemplates/BlazorTemplates.Tests/playwrightSettings.json:
```diff
@@ -4,18 +4,12 @@
-      "Path": "har"
-    },
+    "RecordVideoDir": "videos",
+    "RecordHarPath": "har",
```
I can't make a comment in line 25 (github limitation) but you can remove the args with --ignore-certiicate-errors since you already do it in line 19.

mxschmitt commented on eng/helix/content/RunTests/TestRunner.cs:
```diff
@@ -112,8 +113,32 @@ public async Task<bool> InstallPlaywrightAsync()
+                var tempProjectPath = Path.Combine("Options.HELIX_WORKITEM_ROOT", "TempProj.csproj");
+                File.CreateFile(tempProjectPath);
+
+                await ProcessUtil.RunAsync($"playwright",
```
This whole thing can be replaced by: `var exitCode = Microsoft.Playwright.Program.Main(new[] { "install" });`

We'll have an official better API in the future tho to run CLI commands programatically.

TanayParikh committed:
Merge branch 'main' into taparik/upgradePlaywrightSharpTool

TanayParikh committed:
Remove DOTNET_ROLL_FORWARD

TanayParikh committed:
PR Feedback

TanayParikh committed:
Remove temp project workaround.

TanayParikh committed:
Fix package lowerbound issue

TanayParikh committed:
Update Versions.props

TanayParikh commented:
/azp run

TanayParikh commented:
Needs a full re-run given https://github.com/dotnet/aspnetcore/pull/39361.

TanayParikh committed:
Update BrowserManager.cs

TanayParikh commented on src/ProjectTemplates/BlazorTemplates.Tests/BlazorTemplates.Tests.csproj:
```diff
@@ -9,9 +9,6 @@
     <BaseOutputPath />
     <OutputPath />
 
-    <!-- https://github.com/dotnet/aspnetcore/issues/38818 -->
-    <BuildHelixPayload>false</BuildHelixPayload>
-    
```
We may want to leave this bit as is for now given this PR is focused on the Playwright upgrade (tests are still broken - missing nav link component issue).

TanayParikh commented:
Thanks @mxschmitt, that was really helpful!

mxschmitt commented on eng/helix/content/RunTests/TestRunner.cs:
```diff
@@ -112,8 +109,20 @@ public async Task<bool> InstallPlaywrightAsync()
-                await Playwright.InstallAsync(Environment.GetEnvironmentVariable("PLAYWRIGHT_BROWSERS_PATH"), Environment.GetEnvironmentVariable("PLAYWRIGHT_DRIVER_PATH"));
+                Console.WriteLine($"Installing Microsoft.Playwright.CLI");
+
+                await ProcessUtil.RunAsync($"{Options.DotnetRoot}/dotnet",
+                    $"tool install Microsoft.Playwright.CLI --tool-path {Options.HELIX_WORKITEM_ROOT}",
+                    environmentVariables: EnvironmentVariables,
+                    outputDataReceived: Console.WriteLine,
+                    errorDataReceived: Console.Error.WriteLine,
+                    throwOnError: false,
+                    cancellationToken: new CancellationTokenSource(TimeSpan.FromMinutes(2)).Token);
+
+                Console.WriteLine($"Installing Playwright Browsers to {Environment.GetEnvironmentVariable("PLAYWRIGHT_BROWSERS_PATH")}");
+
```
That's not needed for the browser install.

TanayParikh committed:
Remove `Installing Microsoft.Playwright.CLI`

TanayParikh committed:
Merge branch 'main' into taparik/upgradePlaywrightSharpTool

TanayParikh committed:
Disable playwright tests in helix
